src/Features/CSharp/Portable/MetadataAsSource/CSharpMetadataAsSourceService.cs

```diff
@@ -71,55 +73,162 @@ protected override ImmutableArray<AbstractReducer> GetReducers()
                 new CSharpParenthesizedPatternReducer(),
                 new CSharpDefaultExpressionReducer());
 
-        private class FormattingRule : AbstractMetadataFormattingRule
+        /// <summary>
+        /// Adds <c>#nullable enable</c> and <c>#nullable disable</c> annotations to the file as necessary.  Note that
+        /// this does not try to be 100% accurate, but rather it handles the most common cases out there.  Specifically,
+        /// if a file contains any nullable annotated/not-annotated types, then we prefix the file with <c>#nullable
+        /// enable</c>.  Then if we hit any members that explicitly have *oblivious* types, not no annotated or
+        /// non-annotated types, then we switch to <c>#nullable disable</c> for those specific members.
+        /// <para/>
+        /// This is technically innacurate for possible, but very uncommon cases.  For example, if the user's code
+        /// explicitly did something like this:
+        /// 
+        /// <code>
+        /// public void Goo(string goo,
+        ///                 #nullable disable
+        ///                 string bar
+        ///                 #nullable enable
+        ///                 string baz);
+        /// </code>
+        /// 
+        /// Then we would be unable to handle that.  However, this is highly unlikely to happen, and so we accept the
+        /// inaccuracy for the purpose of simplicity and for handling the much more common cases of either the entire
+        /// file being annotated, or the user individually disabling annotations at the member level.
+        /// </summary>
```

Comment:
i discussed the above algorithm with @jcouv .  We think it should meet the needs of like 99%+ of all code.

Reply:
great!